pkg/workflow_execution.go

```diff
@@ -401,6 +361,49 @@ func (c *Client) createWorkflow(namespace string, workflowTemplateID uint64, wor
 	return
 }
 
+func ensureWorkflowRunsOnDedicatedNode(wf *wfv1.Workflow, config SystemConfig) (*wfv1.Workflow, error) {
+	antiAffinityLabelKey := "onepanel.io/reserves-instance-type"
+	nodeSelectorVal := ""
+	addPodAffinity := false
+	for i := range wf.Spec.Templates {
+		template := &wf.Spec.Templates[i]
+		if template.NodeSelector == nil {
+			continue
+		}
+
+		for _, v := range template.NodeSelector {
+			if strings.Contains(v, "{{workflow.") {
+				parts := strings.Split(strings.Replace(v, "}}", "", -1), ".")
+				paramName := parts[len(parts)-1]
+				for _, param := range wf.Spec.Arguments.Parameters {
+					if param.Name == paramName && param.Value != nil {
+						nodeSelectorVal = *param.Value
+						break
+					}
+				}
+			}
+			break
+		}
+		template.Metadata.Labels = map[string]string{antiAffinityLabelKey: nodeSelectorVal}
+		addPodAffinity = true
+	}
+	if addPodAffinity {
+		wf.Spec.Affinity = &corev1.Affinity{
+			PodAntiAffinity: &corev1.PodAntiAffinity{
+				RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+					{LabelSelector: &metav1.LabelSelector{
```

Comment:
@aleksandrmelnikov can you clean this up and move it to next line?

Reply:
Yes, will do

Reply:
Check now?